src/macaron/database/table_definitions.py

```diff
@@ -1,128 +1,538 @@
 # Licensed under the Universal Permissive License v 1.0 as shown at https://oss.oracle.com/licenses/upl/.
 
-"""
-ORM Table definitions used by macaron internally.
-
-For tables associated with checks see base_check.py.
-"""
+"""ORM Table definitions used by macaron internally."""
+import hashlib
+import logging
+import os
+import string
 from datetime import datetime
+from pathlib import Path
+from typing import Optional, Self
 
-from sqlalchemy import Boolean, ForeignKey, Integer, String
-from sqlalchemy.orm import Mapped, mapped_column
+from packageurl import PackageURL
+from sqlalchemy import Boolean, Column, Enum, ForeignKey, Integer, String, Table, UniqueConstraint
+from sqlalchemy.orm import Mapped, mapped_column, relationship
 
 from macaron.database.database_manager import ORMBase
 from macaron.database.rfc3339_datetime import RFC3339DateTime
+from macaron.errors import CUEExpectationError, CUERuntimeError
+from macaron.slsa_analyzer.provenance.expectations.cue import cue_validator
+from macaron.slsa_analyzer.provenance.expectations.expectation import Expectation
+from macaron.slsa_analyzer.slsa_req import ReqName
+
+logger: logging.Logger = logging.getLogger(__name__)
 
 # TODO: Use UUIDs as primary keys rather than incremental
 
 ################################################################################
-# Analyzer
-#   - Tables corresponding to an invocation of "macaron analyze"
+# Analysis
+#   - Table corresponding to an invocation of "macaron analyze"
 ################################################################################
 
 
-class SLSARequirement(ORMBase):
-    """Table storing the SLSA requirements a repository satisfies."""
-
-    __tablename__ = "_slsa_requirement"
-    repository: Mapped[int] = mapped_column(Integer, ForeignKey("_repository.id"), primary_key=True)
-    requirement: Mapped[str] = mapped_column(String, primary_key=True)
-    requirement_name: Mapped[str] = mapped_column(String, primary_key=False)
-    feedback: Mapped[str] = mapped_column(String, nullable=True)
-
-
-class RepositoryDependency(ORMBase):
-    """Identifies dependencies between repositories."""
-
-    __tablename__ = "_dependency"
-    dependent_repository: Mapped[int] = mapped_column(Integer, ForeignKey("_repository.id"), primary_key=True)
-    dependency_repository: Mapped[int] = mapped_column(Integer, ForeignKey("_repository.id"), primary_key=True)
-
-
-class AnalysisTable(ORMBase):
+class Analysis(ORMBase):
     """
     ORM Class for the analysis information.
 
-    This information pertains to a single invocation of the macaron tool.
+    This information pertains to a single invocation of Macaron.
     """
 
     __tablename__ = "_analysis"
+
+    #: The primary key.
     id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)  # noqa: A003
+
+    #: The analysis start time.
     analysis_time: Mapped[datetime] = mapped_column(RFC3339DateTime, nullable=False)
```

Comment:
No we can't. `RFC3339DateTime` is a [custom type](https://docs.sqlalchemy.org/en/20/core/custom_types.html#sqlalchemy.types.TypeDecorator) for the database column while `datetime` is a Python type. SQLALchemy uses the type implementation for `RFC3339DateTime` to perform the mapping.

Reply:
According to the documentation [here](https://docs.sqlalchemy.org/en/20/orm/declarative_tables.html#mapped-column-derives-the-datatype-and-nullability-from-the-mapped-annotation) - 

> If the [mapped_column()](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column) construct indicates an explicit type as passed to the [mapped_column.__type](https://docs.sqlalchemy.org/en/20/orm/mapping_api.html#sqlalchemy.orm.mapped_column.params.__type) argument, then the given Python type is disregarded.

Does it means that SQLAlchemy will ignore `datetime` in `Mapped[datetime]` because the datatype `RFC3339DateTime` has been explicitly declared in `mapped_column`?

Reply:
SQLALchemy uses the type implementation for `RFC3339DateTime` to perform the mapping. Specifying `Mapped[datetime]` clarifies what the mapped Python type is more from a documentation point of view.